pkg/services/ngalert/schedule/schedule.go

```diff
@@ -487,10 +509,11 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 				}
 			}()
 		case <-grafanaCtx.Done():
-			// clean up the state only if the reason for stopping the evaluation loop is that the rule was deleted
-			if errors.Is(grafanaCtx.Err(), errRuleDeleted) {
-				clearState()
-			}
+			// We need to use context.Background() here because grafanaCtx is being canceled, and we need a valid context
+			// in clearState function to do DB operations.
+			ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancelFunc()
```

Comment:
Apologies, we might need to remove the cancelFunc I think as annotations are done async, the context will be canceled when this function returns, not when the timeout has expired.

Reply:
Yeah, just tested it! I will remove the `defer`.

```
INFO [01-19|18:02:50] alert rule paused                        logger=ngalert.scheduler rule_uid=VJQYoSTVz org_id=1
INFO [01-19|18:02:50] rules state was reset                    logger=ngalert.state.manager rule_uid=VJQYoSTVz org_id=1 states=1
ERROR[01-19|18:02:50] Error saving alert annotation batch      logger=ngalert.state.historian error="context canceled"
```

Reply:
Wow, it worked well for me. Good catch!

Reply:
Make sure you leave a comment about why you do not call cancel. Someone could come later and "fix" it 😄